investos/portfolio/risk_model/factor_risk.py

```diff
@@ -1,23 +1,184 @@
+        self._recalc_freq = recalc_freq                         #pd.Timedelta(days = recalc_freq)
+
+        self._fos_risk_factor_uuids = fos_risk_factor_uuids
+        self._fos_return_uuid = fos_return_uuid
+
+        self._end_date = end_date
+        self._start_date = start_date
+        self._adjustments = adjustments
+        #self._drop_excluded_assets()
+
+        if self._fos_risk_factor_uuids: 
+            print("=================Barra Risk Model================= ")
+            print("\n1. Initialization")
+            self._pull_fos_risk_factor_dfs()
+            print("\n2. Cross Sectional Regression")
+            self._generate_risk_models()
+            print("\n3. Apply Risk Adjustment")
+            self._apply_risk_model_adjustments()
+        else:
+            self._factor_covariance = factor_covariance
+            self._factor_loadings = factor_loadings
+            self._idiosyncratic_variance = idiosyncratic_variance
+        #self._drop_excluded_assets()
+
+    # Initialization
+    def _pull_fos_risk_factor_dfs(self):
+        # pull from feature hub
+        fos.api_key = os.environ.get("FORECASTOS_API_KEY")                  #.env including my api_key
+        if not fos.api_key:
+            raise ValueError("Feature hub connection failed. Please check the API key is valid.")
+        dataframes = [fos.Feature.get(factor_uuids[factor]).get_df().rename(columns={'value': factor})
+                       for factor in (self._fos_risk_factor_uuids + self._fos_return_uuid)  if factor != 'rbics']
+
+        if dataframes:
+            df_merged = reduce(lambda left, right: pd.merge(left, right, how='left', on=['datetime', 'id']), dataframes)
+        else:
+            raise ValueError("factor loading is empty. Please check the risk factor uuids")
+        # special handling industry factors
+        if 'rbics' in self._fos_risk_factor_uuids:
+            df_industry = pd.get_dummies(fos.Feature.get(factor_uuids['rbics']).get_df().rename(columns={'value': 'industry'}),columns = ['industry'])
+            df_industry.iloc[:,1:] = df_industry.iloc[:,1:].astype(int)     # convert bool to int for industry exposure
+            df_merged = df_merged.merge(df_industry,how='left', on='id')
 
-        self.factor_covariance = factor_covariance
-        self.factor_loadings = factor_loadings
-        self.idiosyncratic_variance = idiosyncratic_variance
+            self._fos_risk_factor_uuids.remove('rbics')
+            self._fos_industry_factors = [col for col in df_industry.columns if col.startswith('industry_')]
+        
+        self._df_loadings =df_merged.replace([np.inf, -np.inf], np.nan).dropna()
+        self._df_loadings = self._df_loadings[(self._df_loadings.datetime >= self._end_date) & (self._df_loadings.datetime <= self._start_date)]
+        return
+    
+    def _generate_risk_models(self, timestamp = 'datetime') -> None:
```

Comment:
Any function that is ~20 or greater lines should be split into new functions. Even +10 is a bit long.